### PR TITLE
Fix image reorientation and cropping bugs in conform workflow (fixes #764)

### LINKED ISSUE #764: Fastsurfer crops the image volumes
## Description

Hi there! Thank you for releasing the library! I'm trying to use the `seg_only` functionality to get the DKT atlas labels, but my segmentations etc. are cropped and way off to the side.  The image is a regular T1w image with 1mm isotropic resolution.

Here is my original image (viewed in ITK SNAP): 

Image

Here is the command I'm running: 

`./run_fastsurfer.sh --sid results --sd /data/sub01/fastsurfer --t1 /data/sub01/aligned_norm.nii.gz  --seg_only --no_cc --no_cereb --no_hypothal --threads 8`

Here is the `orig.mgz` (conformed input - notice the crop)

Image

Here is the `orig.mgz` with segmenation overlaid

Image

## Expected Behavior

There should be no cropping in the final results.

## Environment
 - local install and uv

### PR DESCRIPTION
**Pull request by dkuegler**
Problem Statement
----
This PR addresses two related issues with FastSurfer's image reorientation and conforming pipeline:

**Issue 1 - Incorrect Image Cropping (#764)**: The previous implementation had bugs in how images were cropped during reorientation, leading to misaligned data being processed by downstream models. This caused spatial errors in the segmentation pipeline.

**Issue 2 - Image Center Offset During Orientation Flipping (related to [FreeSurfer #1358](https://github.com/freesurfer/freesurfer/issues/1358))**: When images are flipped to change orientation (e.g., LAS→RAS), the affine matrix must correctly maintain the image center. The original code did not properly account for the MGH format's definition of image center at `Ni/2` instead of `(Ni-1)/2`, causing the image content to shift in world coordinates during reorientation.

Solution
---
This PR introduces a comprehensive refactoring of the image orientation code:
1. New `Reorientation` Class: Encapsulates all orientation transformation logic in a dedicated class that:
   - Manages the vox2vox transformation matrix and metadata
   - Provides clear methods for applying forward and inverse transformations
   - Properly handles image center preservation during arbitrary orientation changes
   - Supports both simple reorientation and spatial resampling with interpolation
2. Fixed Transformation Mathematics:
   - Corrects the affine matrix calculations for flipped orientations
   - Properly computes image center offsets based on the shape and orientation
   - Ensures consistent handling of image metadata across the pipeline
3. Improved Code Structure:
   - Separates concerns between orientation detection and transformation application
   - Removes functions (`to_target_orientation`, `orientation_to_ornts`)
   - Simplifies the calling code in `run_prediction.py` with a clearer API
4. Comprehensive Testing:
   - Adds two new test modules with 20+ test cases covering:
   - World coordinate consistency after reorientation
   - Round-trip conversions (original → target → original)
   - Affine matrix correctness for various orientations
   - Edge cases with anisotropic voxels and different image shapes

Changes
---
- **FastSurferCNN/data_loader/conform.py**: Complete refactoring with new Reorientation class and helper functions
- **FastSurferCNN/run_prediction.py**: Updated to use the new Reorientation API
- **FastSurferCNN/utils/common.py**: Added array_flags context manager for temporary array flag manipulation
- **HypVINN/data_loader/data_utils.py**: Simplified to use refactored utilities
- **'test/image/test_conform_reorient.py**: New tests for conform and reorientation functions
- **test/image/test_orientation_transform.py**: New unit tests for orientation transformation internals

Testing
--
All changes are covered by new comprehensive unit tests that verify:
- ✅ Images are correctly reoriented without data loss or corruption
- ✅ World coordinates remain consistent after transformations
- ✅ The image center is preserved during orientation flipping
- ✅ Round-trip conversions return images to their original state
- ✅ Affine matrices are correct for various orientation combinations

Backward Compatibility
---
⚠️ Breaking Changes:
- `to_target_orientation()` has been removed (use `Reorientation.from_target_orientation()` instead)
- `orientation_to_ornts()` has been removed (internal functionality now in Reorientation class)

Related Issues
---
- Fixes #764 (incorrect image cropping)
- Relates to [FreeSurfer #1358](https://github.com/freesurfer/freesurfer/issues/1358)

### CODE REVIEW

File: FastSurferCNN/data_loader/conform.py [hunk 1]
```diff
@@ -837,11 +861,18 @@ def prepare_mgh_header(
...
-    center = np.asarray(img.shape[:3], dtype=float) / 2.0
-    h1["Pxyz_c"] = img.affine.dot(np.hstack((center, [1.0])))[:3]
+    center = (np.asarray(img.shape[:3], dtype=float) - 1 if FIX_MGH_AFFINE_CALCULATION else 0) / 2.0
+    if FIX_CENTER_NOT_CENTER:
+        # The center is not actually the center, but rather the position of the voxel at Ni/2 (counting at voxel 0)
+        # Therefore, the center changes, if we apply a vox2vox
+        # to get to the true center, move back half a voxel in all directions
+        true_center = center - 0.5 * np.ones((1,3)) @ source_affine[:3,:3]
+        # new image center from true center go half a voxel in all direction of the new affine
+        center = 0.5 * np.ones((1,3)) @ get_affine_from_mgh(h1)[:3,:3] + true_center
```
[the remark this answers is not in the document]
Comment (dkuegler) [reply FastSurferCNN/data_loader/conform.py]: This is incorrect, we want `1,3` to get the correct output shape.

**Comment by dkuegler** [comment]
This PR is already rebased on #804 and should be reviewed after integration of #804.

File: FastSurferCNN/segstats.py [hunk 2]
```diff
@@ -1303,8 +1303,9 @@ def _table_body(file: IO, _dataframe: pd.DataFrame) -> None:
         """Write the volume stats from _dataframe to a file."""
         columns = [col for col in COLUMNS if col in _dataframe.columns]
         fmt = " ".join(_column_format(k) for k in columns)
-        for _index, row in _dataframe.iterrows():
-            data = [row[k] for k in columns]
+        # for some reason DataFrame.iterrows sometimes converts integers into floats
+        for values in _dataframe.itertuples():
+            data = [getattr(values, k) for k in columns]
```
[the remark this answers is not in the document]
Comment (dkuegler) [reply FastSurferCNN/segstats.py]: I had to play around with this quite a bit. I want the named tuple, so name=None is not good, but index=False should be fine.

File: HypVINN/data_loader/data_utils.py [hunk 3]
```diff
@@ -80,19 +41,12 @@ def reorient_img(img, ref_img):
...
-        tornt = np.ones_like(ref_ornt)
-        tornt[:, 0] = ref_ornt[:, 0]
-        img = img.as_reoriented(tornt)
-
-    return img
+    # if the affines are the same, no reorientation is required and we can skip this
+    if np.array_equal(ref_img.affine, img.affine):
+        return img
+    from nibabel.orientations import aff2axcodes
+    target_orientation = "soft " + aff2axcodes(ref_img.affine, AXCODES)
```
[the remark this answers is not in the document]
Comment (dkuegler) [reply HypVINN/data_loader/data_utils.py]: Accepted.

File: CerebNet/data_loader/dataset.py [hunk 4]
```diff
@@ -276,8 +276,11 @@ def __init__(
         }
         # crop the region of interest
         img = crop_transform(self.img_org_data, offsets=self.roi["offsets"], target_shape=self.roi["target_shape"])
+        patch_vox2vox = np.concatenate([np.eye(4)[:3], np.append(self.roi["offsets"], 1)], axis=0)
```
[the remark this answers is not in the document]
Comment (dkuegler) [reply CerebNet/data_loader/dataset.py]: Fixed.

File: FastSurferCNN/data_loader/conform.py [hunk 5]
```diff
@@ -1138,10 +1475,12 @@ def conformed_vox_img_size(
         LOGGER.warning("conform_vox_size is deprecated, replaced by vox_size and will be removed.")
         vox_size = kwargs["conform_vox_size"]
 
+    target_vox_size: Vector3d | None
+    target_img_size: IntVector3d | None
     MAX_VOX_SIZE = 1.0
     MAX_DIMENSION = 256
     # this is similar to mri_convert --conform_min
-    if isinstance(vox_size, str) and (vox_size := vox_size.lower()) in ["min", "auto"]:
+    if isinstance(vox_size, str) and (vox_size := "min") in ["min", "auto"]:
```
[the remark this answers is not in the document]
Comment (dkuegler) [reply FastSurferCNN/data_loader/conform.py]: Bad change, where did this come from? Fixed.

File: FastSurferCNN/data_loader/conform.py [hunk 6]
```diff
@@ -420,14 +716,21 @@ def map_image(
...
+                image_data = np.squeeze(image_data, axis=tuple(range(3, image_data.ndim)))  # ty:ignore[invalid-assignment]
         # if the output has the same number of frames as the input
         elif image_data.shape[3:] == out_shape[3:]:
             # add a frame dimension to vox2vox
             _vox2vox = np.eye(5, dtype=vox2vox.dtype)
             _vox2vox[:3, :3] = vox2vox[:3, :3]
             _vox2vox[3:, 4:] = vox2vox[:3, 3:]
-            vox2vox = _vox2vox
+            vox2vox = np.linalg.inv(_vox2vox)
         else:
```
[the remark this answers is not in the document]
Comment (dkuegler) [reply FastSurferCNN/data_loader/conform.py:735]: Fixed.

File: FastSurferCNN/data_loader/conform.py [hunk 7]
```diff
@@ -1339,33 +1675,43 @@ def _crop_transform_pad_fn(image, pad_tuples, pad):
...
-    elif isinstance(image, np.ndarray):
-        kwargs["constant_values"] = pad
-    else:  # Tensor
-        kwargs["value"] = pad
-
-    from functools import partial
-
+        TorchPadModes = ("reflect", "replicate", "circular")
+        NumpyPadModes = ("edge", "linear_ramp", "maximum", "mean", "median", "minimum", "reflect", "symmetric", "wrap")
+        if isinstance(image, np.ndarray) and pad not in NumpyPadModes or pad not in TorchPadModes:
```
[the remark this answers is not in the document]
Comment (dkuegler) [reply FastSurferCNN/data_loader/conform.py]: Fixed.

File: FastSurferCNN/data_loader/conform.py [hunk 8]
```diff
@@ -633,10 +949,10 @@ def conform(
...
         img_size: ImageSizeOption | None = 256,
-        dtype: type | None = np.uint8,
+        dtype: npt.DTypeLike | None = np.uint8,
         orientation: OrientationType | None = "lia",
         threshold_1mm: float | None = None,
-        rescale: int | float | Literal["none"] = 255,
+        rescale: int | float | None = 255,
         vox_eps: float = 1e-4,
         rot_eps: float = 1e-6,
         **kwargs,
```
[the remark this answers is not in the document]
Comment (dkuegler) [reply FastSurferCNN/data_loader/conform.py:960]: Fixed to actually use the file_type parameter.

File: FastSurferCNN/segstats.py [hunk 9]
```diff
@@ -807,7 +807,7 @@ def main(args: argparse.Namespace) -> Literal[0] | str:
             except FileNotFoundError:
                 return f"Could not find the ColorLUT in {lut_file}, make sure the --lut argument is valid."
             except Exception as exception:
-                return exception.args[0]
+                return f"ERROR: Loading the ColorLUT failed with error: {exception.args[0]}"
```
[the remark this answers is not in the document]
Comment (dkuegler) [reply FastSurferCNN/segstats.py]: Changed.
